contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -534,14 +532,22 @@ class USDScene::IO : public RefCounted
 			return m_allTags;
 		}
 
-		pxr::UsdShadeMaterial computeBoundMaterial( const pxr::UsdPrim &prim )
+		pxr::UsdShadeMaterial computeBoundMaterial( const pxr::UsdPrim &prim, const pxr::TfToken &materialPurpose )
 		{
 			// This should be thread safe, despite using caches, because
 			// BindingsCache and CollectionQueryCache are implemented by USD as
 			// tbb::concurrent_unordered_map
-			return pxr::UsdShadeMaterialBindingAPI( prim ).ComputeBoundMaterial(
-				&m_usdBindingsCache, &m_usdCollectionQueryCache
+			pxr::UsdRelationship bindingRelationship;
+			pxr::UsdShadeMaterial material = pxr::UsdShadeMaterialBindingAPI( prim ).ComputeBoundMaterial(
+				&m_usdBindingsCaches.at( materialPurpose ), &m_usdCollectionQueryCache, materialPurpose, &bindingRelationship
 			);
+			if( material && materialPurpose != pxr::UsdShadeTokens->allPurpose && bindingRelationship.GetBaseName() != materialPurpose )
```

Comment:
Some of these clauses seem unnecessary - if the first clause is false, then it doesn't matter whether or not we take this branch, we're ending up with an uninitialized material either way.  And I don't get the second clause - if the purpose we're looking for is allPurpose, then it's not a problem if it falls back to allPurpose ... but won't `bindingRelationship.GetBaseName() == materialPurpose` in that case anyway?  Could we just use the final clause?

Reply:
They do each have a role to play actually, and the tests fail if any one is removed :

- `material` needs to be checked because if no material was found, `bindingRelationship.GetBaseName()` will throw because it hasn't been initialised.
- `materialPurpose != pxr::UsdShadeTokens->allPurpose` needs to be checked, because for all purpose bindings, there is no explicit purpose in the name of the relationship, and `GetBaseName()` will return `"binding"` and not the purpose.

Reply:
Hopefully that makes sense - I'm going to take the liberty of merging so that I can get a new dependencies package out today.

Reply:
Yep, that makes sense.